cli/azd/internal/telemetry/appinsights-exporter/span_to_envelope.go

```diff
@@ -0,0 +1,145 @@
+
+	requestData, contextTags := spanToRequestData(span)
+	data := contracts.NewData()
+	data.BaseData = requestData
+	data.BaseType = requestData.BaseType()
+
+	envelope.Data = data
+	envelope.Name = requestData.EnvelopeName("")
+	envelope.Tags[contracts.OperationName] = requestData.Name
+
+	for contextKey, contextVal := range contextTags {
+		envelope.Tags[contextKey] = contextVal
+	}
+
+	// Sanitize.
+	for _, warn := range envelope.Sanitize() {
+		diagLog.Printf("Telemetry data warning: %s", warn)
+	}
+
+	for _, warn := range contracts.SanitizeTags(envelope.Tags) {
+		diagLog.Printf("Telemetry tag warning: %s", warn)
+	}
+
+	return envelope
+}
+
+// Converts an OpenTelemetry span into an ApplicationInsights RequestTelemetry data,
+// returning remainder attributes that are meant to used as context tags.
+func spanToRequestData(span trace.ReadOnlySpan) (*contracts.RequestData, map[string]string) {
+	contextTags := map[string]string{}
+	requestData := contracts.NewRequestData()
+	requestData.Id = span.SpanContext().SpanID().String()
+	requestData.Name = span.Name()
+	requestData.Source = ""
+	requestData.Duration = formatDuration(span.EndTime().Sub(span.StartTime()))
+	requestData.Success = span.Status().Code != codes.Error
+	requestData.Url = ""
+	requestData.Properties = map[string]string{}
+	requestData.Measurements = map[string]float64{}
+
+	if span.Status().Code == codes.Error {
+		if len(span.Status().Description) > 0 {
+			requestData.ResponseCode = span.Status().Description
+		} else {
+			requestData.ResponseCode = "UnknownFailure"
+		}
+	} else {
+		requestData.ResponseCode = "Success"
+	}
+
+	for _, resourceAttrib := range span.Resource().Attributes() {
+		if isResourceContextTagField(string(resourceAttrib.Key)) {
+			contextTags[string(resourceAttrib.Key)] = resourceAttrib.Value.Emit()
+		} else {
+			SetAttributeAsPropertyOrMeasurement(resourceAttrib, requestData.Properties, requestData.Measurements)
+		}
+	}
+
+	for _, attrib := range span.Attributes() {
+		if isAttributeContextTagField(string(attrib.Key)) {
+			contextTags[string(attrib.Key)] = attrib.Value.Emit()
+		} else {
+			SetAttributeAsPropertyOrMeasurement(attrib, requestData.Properties, requestData.Measurements)
+		}
+	}
+
+	for _, warn := range requestData.Sanitize() {
+		diagLog.Printf("Request telemetry data warning: %s", warn)
+	}
+
+	return requestData, contextTags
+}
+
+func isResourceContextTagField(key string) bool {
+	return key == contracts.ApplicationVersion
+}
+
+var attributeContextTagKeys = map[string]bool{
+	// Currently, we only support user and session related info.
+	// The full list of context tag keys can be found in the appinsights-go library.
+	contracts.UserAccountId:  true,
+	contracts.UserAuthUserId: true,
+	contracts.UserId:         true,
+	contracts.SessionId:      true,
+	contracts.SessionIsFirst: true,
+}
+
+func isAttributeContextTagField(key string) bool {
+	return attributeContextTagKeys[key]
+}
```

Comment:
Since all values are `true`, I wonder if you can just use a `set` instead of the dictionary.
Then, you would just return `true` if the `key` is in the `set` or false if not...  Then we can save all those `true` labels on the map declaration/initialization

Reply:
I don't think there's a `set` as part of the `golang` standard library.. This seems to be the general recommended way..

Reply:
I've seen people use `map[string]struct{}`  to implement a Set. This way the value is zero bytes.

Reply:
The memory optimization is interesting, but immediately for me, wasn't worth considering given the size of the dictionary here. I do however see that storing `map[string]struct{}` prevents a tertiary state of an item with `false` returning the same value as an item not existing. Overall, I think that improves readability.

Going forward I'd lean `map[string]struct{}` as a result. However, the difference is small enough that I'll leave it as-is for this change.